File: _legacy/packages/cli/src/commands/ask.ts
import type { Command } from 'commander';
import chalk from 'chalk';
import {
  isConfigured,
  loadConfig,
  getDb,
  askQuestion,
  type AskResult,
} from '@repog/core';

/**
 * Options passed from command line.
 */
interface AskCommandOptions {
  repo?: string;
  limit?: number;
}

/**
 * Check if the chunk_embeddings table has at least one row.
 */
function hasEmbeddings(): boolean {
  try {
    const config = loadConfig();
    const db = getDb(config.dbPath);
    const row = db.prepare('SELECT COUNT(*) as count FROM chunk_embeddings').get() as {
      count: number;
    };
    return row.count > 0;
  } catch {
    return false;
  }
}

/**
 * Validate repo format is owner/repo.
 */
function isValidRepoFormat(repo: string): boolean {
  const parts = repo.split('/');
  return parts.length === 2 && parts[0].length > 0 && parts[1].length > 0;
}

/**
 * Format similarity as a percentage.
 */
function formatSimilarity(similarity: number): string {
  return `${Math.round(similarity * 100)}%`;
}

/**
 * Render the question header.
 */
function renderHeader(question: string, repo?: string): void {
  console.log(`Question: "${question}"`);
  console.log(chalk.gray('\u2500'.repeat(45)));
  console.log('');

  if (repo) {
    console.log(`Repo: ${repo}`);
    console.log(chalk.gray('\u2500'.repeat(45)));
    console.log('');
  }
}

/**
 * Render the source attribution footer.
 */
function renderFooter(result: AskResult): void {
  console.log('');
  console.log('');
  console.log(chalk.gray('\u2500'.repeat(45)));
  console.log('Sources:');

  for (const source of result.sources) {
    const similarity = formatSimilarity(source.similarity);
    console.log(chalk.dim(`  \u00b7 ${source.repoFullName} (${source.chunkType})  ${similarity}`));
  }

  console.log('');
  console.log(
    chalk.dim(
      `${result.inputTokens} tokens in  \u00b7  ${result.outputTokens} tokens out  \u00b7  ${Math.round(result.durationMs)}ms`
    )
  );
}

/**
 * Run the ask command.
 */
async function runAsk(question: string, options: AskCommandOptions): Promise<void> {
  // Check if configured
  if (!isConfigured()) {
    console.log(chalk.red('Run `repog init` first.'));
    process.exit(1);
  }

  // Check if embeddings exist
  if (!hasEmbeddings()) {
    console.log(chalk.red('No embeddings found. Run `repog sync` then `repog embed` first.'));
    process.exit(1);
  }

  // Validate repo format if provided
  if (options.repo && !isValidRepoFormat(options.repo)) {
    console.log(chalk.red('Invalid repo format. Use owner/repo (e.g. torvalds/linux).'));
    process.exit(1);
  }

  // Handle Ctrl+C gracefully
  const handleSignal = (): void => {
    console.log(''); // Newline to clean up mid-line streaming
    console.log(chalk.dim('Cancelled.'));
    process.exit(0);
  };

  process.on('SIGINT', handleSignal);
  process.on('SIGTERM', handleSignal);

  // Render header before starting stream
  renderHeader(question, options.repo);

  try {
    // Stream callback - write directly to stdout
    const onChunk = (text: string): void => {
      process.stdout.write(text);
    };

    const result = await askQuestion(
      {
        question,
        repo: options.repo,
        limit: options.limit ?? 10,
      },
      onChunk
    );

    // Remove signal handlers after completion
    process.off('SIGINT', handleSignal);
    process.off('SIGTERM', handleSignal);

    // Check if this is an empty result (no sources)
    if (result.sources.length === 0) {
      // The answer was already streamed, but we need special handling
      // If the answer indicates no information found, show the empty state
      if (
        result.answer.includes("couldn't find any relevant information") ||
        result.inputTokens === 0
      ) {
        // Clear what we've written and show empty state
        // Actually the answer was already streamed via onChunk, so just add the hint
        console.log('');
        console.log('');
        console.log(chalk.dim('Try running `repog sync` and `repog embed` to refresh your knowledge base.'));
      } else {
        // Error case - already printed via streaming
        console.log('');
      }
    } else {
      // Normal case - render footer with sources
      renderFooter(result);
    }
  } catch (error) {
    // Remove signal handlers on error
    process.off('SIGINT', handleSignal);
    process.off('SIGTERM', handleSignal);

    console.log('');
    if (error instanceof Error) {
      console.error(chalk.red(`Error: ${error.message}`));
    } else {
      console.error(chalk.red('An unknown error occurred.'));
    }
    process.exit(1);
  }
}

/**
 * Register the `repog ask` command.
 * Ask questions about your repositories using RAG.
 */
export function register(program: Command): void {
  program
    .command('ask <question>')
    .description('Ask a question about your repositories using AI')
    .option('-r, --repo <name>', 'Scope question to a specific repository (owner/repo)')
    .option('-l, --limit <number>', 'Number of context chunks to retrieve', (v) => parseInt(v, 10), 10)
    .action(async (question: string, options: AskCommandOptions) => {
      await runAsk(question, options);
    });
}

File: commands/ask.go
package commands

import (
	"context"
	"fmt"
	"os"
	"strings"

	"github.com/fatih/color"
	"github.com/spf13/cobra"

	"github.com/hackastak/repog/internal/ask"
	"github.com/hackastak/repog/internal/config"
	"github.com/hackastak/repog/internal/db"
	"github.com/hackastak/repog/internal/format"
	"github.com/hackastak/repog/internal/provider"
	_ "github.com/hackastak/repog/internal/provider/gemini"
)

var askCmd = &cobra.Command{
	Use:   "ask <question>",
	Short: "Ask a question about your repositories",
	Long:  "Use RAG to answer questions about your repository knowledge base.",
	Args:  cobra.ExactArgs(1),
	RunE:  runAsk,
}

var (
	askRepo  string
	askLimit int
)

func init() {
	askCmd.Flags().StringVar(&askRepo, "repo", "", "Scope to a specific repository (owner/repo)")
	askCmd.Flags().IntVar(&askLimit, "limit", 10, "Number of chunks to retrieve as context")
}

func runAsk(cmd *cobra.Command, args []string) error {
	question := args[0]

	red := color.New(color.FgRed).SprintFunc()
	cyan := color.New(color.FgCyan).SprintFunc()
	dim := color.New(color.Faint).SprintFunc()
	bold := color.New(color.Bold).SprintFunc()

	// Validate repo format if provided
	if askRepo != "" && !strings.Contains(askRepo, "/") {
		fmt.Println(red("Invalid repository format. Use owner/repo (e.g., hackastak/repog)"))
		os.Exit(1)
	}

	// Load config
	cfg, err := config.LoadConfig()
	if err != nil {
		fmt.Println(red("Run `repog init` first."))
		os.Exit(1)
	}

	// Create embedding provider
	embedKey, err := config.GetAPIKeyForProvider(cfg.Embedding.Provider)
	if err != nil {
		fmt.Println(red("Failed to get embedding API key:"), err)
		os.Exit(1)
	}

	embedProvider, err := provider.NewEmbeddingProvider(cfg.Embedding, embedKey)
	if err != nil {
		fmt.Println(red("Failed to create embedding provider:"), err)
		os.Exit(1)
	}

	// Create LLM provider
	genKey, err := config.GetAPIKeyForProvider(cfg.Generation.Provider)
	if err != nil {
		fmt.Println(red("Failed to get generation API key:"), err)
		os.Exit(1)
	}

	llmProvider, err := provider.NewLLMProvider(cfg.Generation, genKey)
	if err != nil {
		fmt.Println(red("Failed to create LLM provider:"), err)
		os.Exit(1)
	}

	// Open database
	database, err := db.Open(cfg.DBPath, cfg.Embedding.Dimensions)
	if err != nil {
		fmt.Println(red("Database error:"), err)
		os.Exit(1)
	}
	defer func() { _ = database.Close() }()

	// Check if embeddings exist
	var count int
	if err := database.QueryRow("SELECT COUNT(*) FROM chunk_embeddings").Scan(&count); err != nil {
		fmt.Println(red("Database error:"), err)
		os.Exit(1)
	}
	if count == 0 {
		fmt.Println(red("No embeddings found. Run `repog sync` then `repog embed` first."))
		os.Exit(1)
	}

	// Print question header
	fmt.Println()
	fmt.Println(bold("Question:"), question)
	fmt.Println()
	fmt.Print(bold("Answer: "))

	// Stream answer
	result, err := ask.AskQuestion(context.Background(), ask.AskOptions{
		Question:          question,
		Repo:              askRepo,
		Limit:             askLimit,
		DB:                database,
		EmbeddingProvider: embedProvider,
		LLMProvider:       llmProvider,
	}, func(chunk string) {
		fmt.Print(chunk)
	})

	if err != nil {
		fmt.Println()
		fmt.Println(red("Error:"), err)
		os.Exit(1)
	}

	fmt.Println() // End the streaming line
	fmt.Println()

	// Print sources
	if len(result.Sources) > 0 {
		fmt.Println(dim("Sources:"))
		for _, source := range result.Sources {
			similarity := format.FormatSimilarity(source.Similarity)
			fmt.Println(" ", cyan(source.RepoFullName), dim("("+source.ChunkType+")"), dim(similarity))
		}
		fmt.Println()
	}

	// Print metrics
	fmt.Println(dim(fmt.Sprintf("Tokens: %d in / %d out | Duration: %dms",
		result.InputTokens, result.OutputTokens, result.DurationMs)))

	return nil
}
